lighthouse-core/gather/driver/target-manager.js

```diff
@@ -0,0 +1,125 @@
+/**
+ * @license Copyright 2021 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/**
+  * @fileoverview This class tracks multiple targets (the page itself and its OOPIFs) and allows consumers to
+  * listen for protocol events before each target is resumed.
+  */
+
+const log = require('lighthouse-logger');
+
+/** @typedef {{target: LH.Crdp.Target.TargetInfo, session: LH.Gatherer.FRProtocolSession}} TargetWithSession */
+
+class TargetManager {
+  /** @param {LH.Gatherer.FRProtocolSession} session */
+  constructor(session) {
+    this._enabled = false;
+    this._session = session;
+    /** @type {Array<(targetWithSession: TargetWithSession) => Promise<void>|void>} */
+    this._listeners = [];
+
+    this._onSessionAttached = this._onSessionAttached.bind(this);
+
+    /** @type {Map<string, TargetWithSession>} */
+    this._targets = new Map();
+
+    /** @param {LH.Crdp.Page.FrameNavigatedEvent} event */
+    this._onFrameNavigated = async event => {
+      // Child frames are handled in `_onSessionAttached`.
+      if (event.frame.parentId) return;
+
+      // It's not entirely clear when this is necessary, but when the page switches processes on
+      // navigating from about:blank to the `requestedUrl`, resetting `setAutoAttach` has been
+      // necessary in the past.
+      await this._session.sendCommand('Target.setAutoAttach', {
+        autoAttach: true,
+        flatten: true,
+        waitForDebuggerOnStart: true,
+      });
+    };
+  }
+
+  /**
+    * @param {LH.Gatherer.FRProtocolSession} session
+    */
+  async _onSessionAttached(session) {
+    try {
+      const target = await session.sendCommand('Target.getTargetInfo').catch(() => null);
+      const targetType = target && target.targetInfo && target.targetInfo.type;
+      const hasValidTargetType = targetType === 'page' || targetType === 'iframe';
+      if (!target || !hasValidTargetType) return;
+
+      const targetId = target.targetInfo.targetId;
+      session.setTargetInfo(target.targetInfo);
+      if (this._targets.has(targetId)) return;
+
+      const targetName = target.targetInfo.url || target.targetInfo.targetId;
+      log.verbose('target-manager', `target ${targetName} attached`);
+
+      const targetWithSession = {target: target.targetInfo, session};
+      this._targets.set(targetId, targetWithSession);
```

Comment:
is `targetWithSession` in this map eventually needed or could it just be a `Set` of targetIds?

Reply:
and is the plan still to eventually have multiple sessions per target?

Reply:
> and is the plan still to eventually have multiple sessions per target?

yep

> is targetWithSession in this map eventually needed or could it just be a Set of targetIds?

I toyed around with versions that exposed the targets to consumers which is why they're here, and I think we'll eventually do that for creating new sessions/exploring targets, but if it's concerning we can switch to just the set.

Reply:
> I toyed around with versions that exposed the targets to consumers which is why they're here, and I think we'll eventually do that for creating new sessions/exploring targets, but if it's concerning we can switch to just the set.

just live blogging me reading through the code and reasoning through what things are used for what. `Set` is probably the right thing to use for the current code, but it's not like there's going to be some deep YAGNI penalty for using a map with unused (for now) values over a set :)